### DISCUSSION TITLE
Support joining input schemas

### DISCUSSION THREAD

**Discussion by laurent512** [next-safe-action/next-safe-action#173] [Ideas]
Hi, 

Thanks for the good work :) 
Related to the V7, is there a way to chain schema so that they can be deconstructed in the action like the following ? 

```typescript
export const testAction = authActionClient
  .schema(z.object({ key: z.string() }))
  .schema(z.object({ id: z.string() }))
  .action(async ({ parsedInput: { key, id }, ctx: { userId } }) => {
    return {
      updated: true,
    };
  });
```

The reason I ask that is to have part of the schema defined at the client level :)

**Comment by laurent512** [comment]
For extra information, I tried, but it only allows you to deconstruct the last schema content, so here I would only have access to "id"

**Reply by TheEdoRan** [reply]
Yes, this is expected. What if you pass a non-object schema to the first or second `schema` method? Should the library overwrite it or throw a runtime error?
In addition, next-safe-action supports multiple validation libraries via TypeSchema, so it's all abstracted under the hood, and we can't just directly use methods from Zod (or any other lib) to extend object schemas in the library code.
A possible solution would be to pass the previous schema as a parameter to the[ schema callback function](https://next-safe-action.dev/docs/recipes/i18n) where you can then extend it using the proper method from the validation library of your choice (for instance, using [`extend`](https://zod.dev/?id=extend) with Zod).

**Reply by laurent512** [reply]
Thank you for your reply. 

Would it be possible to have an optional argument in the schema method that would define what is done with the following schema provided if any? 

```
function multipleschemahandler(current,next){
return ...
}

 export const testAction = authActionClient
        .schema(z.object({ key: z.string() }) , multipleschemahandler)
        .schema(z.object({ id: z.string() }))
        .action(async ({ parsedInput: { key, id }, ctx: { userId } }) => {
                      return {
                updated: true,
            };
        });
```

**Reply by TheEdoRan** [reply]
Don't think that would work. Instead, as explained above, the previous schema could be passed to the schema function, if you need to access it, like this:

```typescript
export const testAction = authActionClient
  .schema(z.object({ key: z.string() }))
  .schema(async (prevSchema) => prevSchema.extend({ id: z.string() }))
  .action(async ({ parsedInput: { key, id }, ctx: { userId } }) => {
    return {
      updated: true,
    };
  });
```

I think this is doable and clean.

**Reply by laurent512** [reply]
![image](https://github.com/TheEdoRan/next-safe-action/assets/4173291/80be2247-3bb7-43b8-a868-5111e216aba5)

Ok I think it should work for my case, thank you :) 
Currently it gives a typescript error, is it expected?

**Reply by TheEdoRan** [reply]
Yes, it's not implemented yet. I'll work on this feature soon.

**Comment by TheEdoRan** [comment]
Added support for this feature in [v7.1.0](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.1.0), recipe [here](https://next-safe-action.dev/docs/recipes/extend-previous-schema).

**Reply by laurent512** [reply]
thank you <3

**Comment by laurent512** [comment]
By the way, I think it would be nice to have a way to define the schema joining method at the root of the safe actions instead. 
That would simplify the generation of many actions based on the same action client. 
( client = 1 schema joining method and 1 schema ;   action = 1 schema)
It means that it does not rely on the creator of the schema to ensure the correct inheritance. 

You said that this is impossible?
